### PR TITLE
Initialize operation arguments with ONNX model constants

### PR DESCRIPTION
**Pull request by doru1004**
This patch enables the emission of constant values to be passed to operations. The current patch applies this functionality to the second argument of the Reshape operation.

The outcome is the shape dimensions for Reshape part of the MNIST model are now known:

```
module {
  func @main_graph(%arg0: tensor<1x1x28x28xf32>, %arg1: tensor<8x1x5x5xf32>, %arg2: tensor<8x1x1xf32>, %arg3: tensor<16x8x5x5xf32>, %arg4: tensor<16x1x1xf32>, %arg5: tensor<2xi64>, %arg6: tensor<16x4x4x10xf32>, %arg7: tensor<2xi64>, %arg8: tensor<1x10xf32>) -> tensor<1x10xf32> {
    %0 = "onnx.ConstantTensor"() {value = [256, 10]} : () -> tensor<2xi64>
    %1 = "onnx.Reshape"(%arg6, %0) : (tensor<16x4x4x10xf32>, tensor<2xi64>) -> tensor<256x10xf32>
    %2 = "onnx.ConvNoBias"(%arg0, %arg1) {auto_pad = "SAME_UPPER", dilations = [1, 1], group = 1 : i64, kernel_shape = [5, 5], strides = [1, 1]} : (tensor<1x1x28x28xf32>, tensor<8x1x5x5xf32>) -> tensor<1x8x28x28xf32>
    %3 = "onnx.Add"(%2, %arg2) : (tensor<1x8x28x28xf32>, tensor<8x1x1xf32>) -> tensor<1x8x28x28xf32>
    %4 = "onnx.Relu"(%3) : (tensor<1x8x28x28xf32>) -> tensor<1x8x28x28xf32>
    %5 = "onnx.MaxPoolSingleOut"(%4) {auto_pad = "NOTSET", dilations = [1, 1], kernel_shape = [2, 2], pads = [0, 0, 0, 0], strides = [2, 2]} : (tensor<1x8x28x28xf32>) -> tensor<1x8x14x14xf32>
    %6 = "onnx.ConvNoBias"(%5, %arg3) {auto_pad = "SAME_UPPER", dilations = [1, 1], group = 1 : i64, kernel_shape = [5, 5], strides = [1, 1]} : (tensor<1x8x14x14xf32>, tensor<16x8x5x5xf32>) -> tensor<1x16x14x14xf32>
    %7 = "onnx.Add"(%6, %arg4) : (tensor<1x16x14x14xf32>, tensor<16x1x1xf32>) -> tensor<1x16x14x14xf32>
    %8 = "onnx.Relu"(%7) : (tensor<1x16x14x14xf32>) -> tensor<1x16x14x14xf32>
    %9 = "onnx.MaxPoolSingleOut"(%8) {auto_pad = "NOTSET", dilations = [1, 1], kernel_shape = [3, 3], pads = [0, 0, 0, 0], strides = [3, 3]} : (tensor<1x16x14x14xf32>) -> tensor<1x16x4x4xf32>
    %10 = "onnx.ConstantTensor"() {value = [1, 256]} : () -> tensor<2xi64>
    %11 = "onnx.Reshape"(%9, %10) : (tensor<1x16x4x4xf32>, tensor<2xi64>) -> tensor<1x256xf32>
    %12 = "onnx.Gemm"(%11, %1, %arg8) {alpha = 1.000000e+00 : f32, beta = 1.000000e+00 : f32, transA = 0 : i64, transB = 0 : i64} : (tensor<1x256xf32>, tensor<256x10xf32>, tensor<1x10xf32>) -> tensor<1x10xf32>
    return %12 : tensor<1x10xf32>
  }
  "onnx.EntryPoint"() {func = @main_graph, numInputs = 1 : i32, numOutputs = 1 : i32} : () -> ()
}
```

Subsequently we lower the above model to the KRNL dialect:

```
module {
  func @main_graph(%arg0: memref<1x1x28x28xf32>, %arg1: memref<8x1x5x5xf32>, %arg2: memref<8x1x1xf32>, %arg3: memref<16x8x5x5xf32>, %arg4: memref<16x1x1xf32>, %arg5: memref<2xi64>, %arg6: memref<16x4x4x10xf32>, %arg7: memref<2xi64>, %arg8: memref<1x10xf32>) -> memref<1x10xf32> {
    %0 = alloc() : memref<1x10xf32>
    %1 = alloc() : memref<1x256xf32>
    %2 = alloc() : memref<1x16x4x4xf32>
    %3 = alloc() : memref<1x16x14x14xf32>
    %4 = alloc() : memref<1x16x14x14xf32>
    %5 = alloc() : memref<1x16x14x14xf32>
    %6 = alloc() : memref<1x8x14x14xf32>
    %7 = alloc() : memref<1x8x28x28xf32>
    %8 = alloc() : memref<1x8x28x28xf32>
    %9 = alloc() : memref<1x8x28x28xf32>
    %10 = alloc() : memref<256x10xf32>
    %c4_i64 = constant 4 : i64
    %c16_i64 = constant 16 : i64
    %11 = muli %c4_i64, %c16_i64 : i64
    %c4_i64_0 = constant 4 : i64
    %12 = muli %11, %c4_i64_0 : i64
    %c4_i64_1 = constant 4 : i64
    %13 = muli %12, %c4_i64_1 : i64
    %c10_i64 = constant 10 : i64
    %14 = muli %13, %c10_i64 : i64
    "krnl.memcpy"(%10, %arg6, %14) : (memref<256x10xf32>, memref<16x4x4x10xf32>, i64) -> ()
    %c8 = constant 8 : index
    %cst = constant 0.000000e+00 : f32
    %c1 = constant 1 : index
    %15:2 = krnl.define_loops 2
    %16:2 = krnl.optimize_loops  {
      krnl.return_loops %15#0, %15#1
    } : () -> (!krnl.loop, !krnl.loop)
    krnl.iterate(%16#0, %16#1) with (%15#0 -> %arg9 = 0 to 1, %15#1 -> %arg10 = 0 to 8) {
      %37:2 = krnl.define_loops 2
      %38:2 = krnl.optimize_loops  {
        krnl.return_loops %37#0, %37#1
      } : () -> (!krnl.loop, !krnl.loop)
      krnl.iterate(%38#0, %38#1) with (%37#0 -> %arg11 = 0 to 28, %37#1 -> %arg12 = 0 to 28) {
        store %cst, %9[%arg9, %arg10, %arg11, %arg12] : memref<1x8x28x28xf32>
        %39:3 = krnl.define_loops 3
        %40:3 = krnl.optimize_loops  {
          krnl.return_loops %39#0, %39#1, %39#2
        } : () -> (!krnl.loop, !krnl.loop, !krnl.loop)
        krnl.iterate(%40#0, %40#1, %40#2) with (%39#0 -> %arg13 = 0 to 1, %39#1 -> %arg14 = 0 to 5, %39#2 -> %arg15 = 0 to 5) {
          %41 = addi %arg11, %arg14 : index
          %42 = addi %arg12, %arg15 : index
          %43 = load %arg0[%arg9, %arg13, %41, %42] : memref<1x1x28x28xf32>
          %44 = load %arg1[%arg10, %arg13, %arg14, %arg15] : memref<8x1x5x5xf32>
          %45 = load %9[%arg9, %arg10, %arg11, %arg12] : memref<1x8x28x28xf32>
          %46 = mulf %43, %44 : f32
          %47 = addf %45, %46 : f32
          store %47, %9[%arg9, %arg10, %arg11, %arg12] : memref<1x8x28x28xf32>
        }
      }
    }
    %17:4 = krnl.define_loops 4
    %18:4 = krnl.optimize_loops  {
      krnl.return_loops %17#0, %17#1, %17#2, %17#3
    } : () -> (!krnl.loop, !krnl.loop, !krnl.loop, !krnl.loop)
    krnl.iterate(%18#0, %18#1, %18#2, %18#3) with (%17#0 -> %arg9 = 0 to 1, %17#1 -> %arg10 = 0 to 8, %17#2 -> %arg11 = 0 to 28, %17#3 -> %arg12 = 0 to 28) {
      %c0 = constant 0 : index
      %37 = load %9[%c0, %arg10, %arg11, %arg12] : memref<1x8x28x28xf32>
      %c0_10 = constant 0 : index
      %c0_11 = constant 0 : index
      %38 = load %arg2[%arg10, %c0_11, %c0_10] : memref<8x1x1xf32>
      %39 = addf %37, %38 : f32
      store %39, %8[%arg9, %arg10, %arg11, %arg12] : memref<1x8x28x28xf32>
    }
    %19:4 = krnl.define_loops 4
    %20:4 = krnl.optimize_loops  {
      krnl.return_loops %19#0, %19#1, %19#2, %19#3
    } : () -> (!krnl.loop, !krnl.loop, !krnl.loop, !krnl.loop)
    krnl.iterate(%20#0, %20#1, %20#2, %20#3) with (%19#0 -> %arg9 = 0 to 1, %19#1 -> %arg10 = 0 to 8, %19#2 -> %arg11 = 0 to 28, %19#3 -> %arg12 = 0 to 28) {
      %37 = load %8[%arg9, %arg10, %arg11, %arg12] : memref<1x8x28x28xf32>
      %cst_10 = constant 0.000000e+00 : f32
      %38 = cmpf "olt", %37, %cst_10 : f32
      %39 = select %38, %cst_10, %37 : f32
      store %39, %7[%arg9, %arg10, %arg11, %arg12] : memref<1x8x28x28xf32>
    }
    %21:4 = krnl.define_loops 4
    %22:4 = krnl.optimize_loops  {
      krnl.return_loops %21#0, %21#1, %21#2, %21#3
    } : () -> (!krnl.loop, !krnl.loop, !krnl.loop, !krnl.loop)
    krnl.iterate(%22#0, %22#1, %22#2, %22#3) with (%21#0 -> %arg9 = 0 to 1, %21#1 -> %arg10 = 0 to 8, %21#2 -> %arg11 = 0 to 14, %21#3 -> %arg12 = 0 to 14) {
      %cst_10 = constant 0xFF800000 : f32
      store %cst_10, %6[%arg9, %arg10, %arg11, %arg12] : memref<1x8x14x14xf32>
      %37:2 = krnl.define_loops 2
      %38:2 = krnl.optimize_loops  {
        krnl.return_loops %37#0, %37#1
      } : () -> (!krnl.loop, !krnl.loop)
      krnl.iterate(%38#0, %38#1) with (%37#0 -> %arg13 = 0 to 2, %37#1 -> %arg14 = 0 to 2) {
        %c2 = constant 2 : index
        %39 = muli %c2, %arg11 : index
        %40 = addi %39, %arg13 : index
        %c2_11 = constant 2 : index
        %41 = muli %c2_11, %arg12 : index
        %42 = addi %41, %arg14 : index
        %43 = load %7[%arg9, %arg10, %40, %42] : memref<1x8x28x28xf32>
        %44 = load %6[%arg9, %arg10, %arg11, %arg12] : memref<1x8x14x14xf32>
        %45 = cmpf "ogt", %44, %43 : f32
        %46 = select %45, %44, %43 : f32
        store %46, %6[%arg9, %arg10, %arg11, %arg12] : memref<1x8x14x14xf32>
      }
    }
    %c16 = constant 16 : index
    %cst_2 = constant 0.000000e+00 : f32
    %c8_3 = constant 8 : index
    %23:2 = krnl.define_loops 2
    %24:2 = krnl.optimize_loops  {
      krnl.return_loops %23#0, %23#1
    } : () -> (!krnl.loop, !krnl.loop)
    krnl.iterate(%24#0, %24#1) with (%23#0 -> %arg9 = 0 to 1, %23#1 -> %arg10 = 0 to 16) {
      %37:2 = krnl.define_loops 2
      %38:2 = krnl.optimize_loops  {
        krnl.return_loops %37#0, %37#1
      } : () -> (!krnl.loop, !krnl.loop)
      krnl.iterate(%38#0, %38#1) with (%37#0 -> %arg11 = 0 to 14, %37#1 -> %arg12 = 0 to 14) {
        store %cst_2, %5[%arg9, %arg10, %arg11, %arg12] : memref<1x16x14x14xf32>
        %39:3 = krnl.define_loops 3
        %40:3 = krnl.optimize_loops  {
          krnl.return_loops %39#0, %39#1, %39#2
        } : () -> (!krnl.loop, !krnl.loop, !krnl.loop)
        krnl.iterate(%40#0, %40#1, %40#2) with (%39#0 -> %arg13 = 0 to 8, %39#1 -> %arg14 = 0 to 5, %39#2 -> %arg15 = 0 to 5) {
          %41 = addi %arg11, %arg14 : index
          %42 = addi %arg12, %arg15 : index
          %43 = load %6[%arg9, %arg13, %41, %42] : memref<1x8x14x14xf32>
          %44 = load %arg3[%arg10, %arg13, %arg14, %arg15] : memref<16x8x5x5xf32>
          %45 = load %5[%arg9, %arg10, %arg11, %arg12] : memref<1x16x14x14xf32>
          %46 = mulf %43, %44 : f32
          %47 = addf %45, %46 : f32
          store %47, %5[%arg9, %arg10, %arg11, %arg12] : memref<1x16x14x14xf32>
        }
      }
    }
    %25:4 = krnl.define_loops 4
    %26:4 = krnl.optimize_loops  {
      krnl.return_loops %25#0, %25#1, %25#2, %25#3
    } : () -> (!krnl.loop, !krnl.loop, !krnl.loop, !krnl.loop)
    krnl.iterate(%26#0, %26#1, %26#2, %26#3) with (%25#0 -> %arg9 = 0 to 1, %25#1 -> %arg10 = 0 to 16, %25#2 -> %arg11 = 0 to 14, %25#3 -> %arg12 = 0 to 14) {
      %c0 = constant 0 : index
      %37 = load %5[%c0, %arg10, %arg11, %arg12] : memref<1x16x14x14xf32>
      %c0_10 = constant 0 : index
      %c0_11 = constant 0 : index
      %38 = load %arg4[%arg10, %c0_11, %c0_10] : memref<16x1x1xf32>
      %39 = addf %37, %38 : f32
      store %39, %4[%arg9, %arg10, %arg11, %arg12] : memref<1x16x14x14xf32>
    }
    %27:4 = krnl.define_loops 4
    %28:4 = krnl.optimize_loops  {
      krnl.return_loops %27#0, %27#1, %27#2, %27#3
    } : () -> (!krnl.loop, !krnl.loop, !krnl.loop, !krnl.loop)
    krnl.iterate(%28#0, %28#1, %28#2, %28#3) with (%27#0 -> %arg9 = 0 to 1, %27#1 -> %arg10 = 0 to 16, %27#2 -> %arg11 = 0 to 14, %27#3 -> %arg12 = 0 to 14) {
      %37 = load %4[%arg9, %arg10, %arg11, %arg12] : memref<1x16x14x14xf32>
      %cst_10 = constant 0.000000e+00 : f32
      %38 = cmpf "olt", %37, %cst_10 : f32
      %39 = select %38, %cst_10, %37 : f32
      store %39, %3[%arg9, %arg10, %arg11, %arg12] : memref<1x16x14x14xf32>
    }
    %29:4 = krnl.define_loops 4
    %30:4 = krnl.optimize_loops  {
      krnl.return_loops %29#0, %29#1, %29#2, %29#3
    } : () -> (!krnl.loop, !krnl.loop, !krnl.loop, !krnl.loop)
    krnl.iterate(%30#0, %30#1, %30#2, %30#3) with (%29#0 -> %arg9 = 0 to 1, %29#1 -> %arg10 = 0 to 16, %29#2 -> %arg11 = 0 to 4, %29#3 -> %arg12 = 0 to 4) {
      %cst_10 = constant 0xFF800000 : f32
      store %cst_10, %2[%arg9, %arg10, %arg11, %arg12] : memref<1x16x4x4xf32>
      %37:2 = krnl.define_loops 2
      %38:2 = krnl.optimize_loops  {
        krnl.return_loops %37#0, %37#1
      } : () -> (!krnl.loop, !krnl.loop)
      krnl.iterate(%38#0, %38#1) with (%37#0 -> %arg13 = 0 to 3, %37#1 -> %arg14 = 0 to 3) {
        %c3 = constant 3 : index
        %39 = muli %c3, %arg11 : index
        %40 = addi %39, %arg13 : index
        %c3_11 = constant 3 : index
        %41 = muli %c3_11, %arg12 : index
        %42 = addi %41, %arg14 : index
        %43 = load %3[%arg9, %arg10, %40, %42] : memref<1x16x14x14xf32>
        %44 = load %2[%arg9, %arg10, %arg11, %arg12] : memref<1x16x4x4xf32>
        %45 = cmpf "ogt", %44, %43 : f32
        %46 = select %45, %44, %43 : f32
        store %46, %2[%arg9, %arg10, %arg11, %arg12] : memref<1x16x4x4xf32>
      }
    }
    %c4_i64_4 = constant 4 : i64
    %c1_i64 = constant 1 : i64
    %31 = muli %c4_i64_4, %c1_i64 : i64
    %c16_i64_5 = constant 16 : i64
    %32 = muli %31, %c16_i64_5 : i64
    %c4_i64_6 = constant 4 : i64
    %33 = muli %32, %c4_i64_6 : i64
    %c4_i64_7 = constant 4 : i64
    %34 = muli %33, %c4_i64_7 : i64
    "krnl.memcpy"(%1, %2, %34) : (memref<1x256xf32>, memref<1x16x4x4xf32>, i64) -> ()
    %cst_8 = constant 1.000000e+00 : f32
    %cst_9 = constant 1.000000e+00 : f32
    %35:3 = krnl.define_loops 3
    %36:3 = krnl.optimize_loops  {
      krnl.return_loops %35#0, %35#1, %35#2
    } : () -> (!krnl.loop, !krnl.loop, !krnl.loop)
    krnl.iterate(%36#0, %36#1) with (%35#0 -> %arg9 = 0 to 1, %35#1 -> %arg10 = 0 to 10) {
      %cst_10 = constant 0.000000e+00 : f32
      store %cst_10, %0[%arg9, %arg10] : memref<1x10xf32>
      krnl.iterate(%36#2) with (%35#2 -> %arg11 = 0 to 256) {
        %42 = load %1[%arg9, %arg11] : memref<1x256xf32>
        %43 = load %10[%arg11, %arg10] : memref<256x10xf32>
        %44 = load %0[%arg9, %arg10] : memref<1x10xf32>
        %45 = mulf %42, %43 : f32
        %46 = addf %44, %45 : f32
        store %46, %0[%arg9, %arg10] : memref<1x10xf32>
      }
      %37 = load %0[%arg9, %arg10] : memref<1x10xf32>
      %38 = mulf %cst_8, %37 : f32
      %c0 = constant 0 : index
      %39 = load %arg8[%c0, %arg10] : memref<1x10xf32>
      %40 = mulf %cst_9, %39 : f32
      %41 = addf %38, %40 : f32
      store %41, %0[%arg9, %arg10] : memref<1x10xf32>
    }
    dealloc %10 : memref<256x10xf32>
    dealloc %9 : memref<1x8x28x28xf32>
    dealloc %8 : memref<1x8x28x28xf32>
    dealloc %7 : memref<1x8x28x28xf32>
    dealloc %6 : memref<1x8x14x14xf32>
    dealloc %5 : memref<1x16x14x14xf32>
    dealloc %4 : memref<1x16x14x14xf32>
    dealloc %3 : memref<1x16x14x14xf32>
    dealloc %2 : memref<1x16x4x4xf32>
    dealloc %1 : memref<1x256xf32>
    return %0 : memref<1x10xf32>
  }
  "krnl.entry_point"() {func = @main_graph, numInputs = 1 : i32, numOutputs = 1 : i32} : () -> ()
}
```

### CODE REVIEW

**Comment by doru1004** [comment]
@tjingrant I think the patch is ready for you to have a look at it. Please let me know your thoughts on it.

**Comment by doru1004** [comment]
> 2\. On how to express the semantic of the calling the model, I prefer to using the existing data flow. The additional mapping is a special case which is difficult to handle. I think we need to clone the model for each call for specialization. For this  test case, we should just check whether the operand is ConstantOp.  It is not a general solution to check the input is initializedTensors in frontend import process. Think about the cascade effect of constant propagation.

I don't think I understand the point you're trying to make here. Can you please elaborate?

**Comment by doru1004** [comment]
Fix to SAME_UPPER and SAME_LOWER auto padding attributes is in: https://github.com/onnx/onnx-mlir/pull/11

**Comment by chentong319** [comment]
I mean to translate the parameter passing into a assignment. For a constant input to arg5, we can have %arg5 = ConstantOp() {tensor_value = [0, 2]} when the function (model) is cloned for a set of parameter. Then  we can use the existing methods to handle it, no need to check the special mapping.

**Comment by chentong319** [comment]
In other words, the solution for reshape with constant second operand should not be in frontend importer.  It should be independent from the solution of parameter passing for model.

**Comment by doru1004** [comment]
> I mean to translate the parameter passing into a assignment. For a constant input to arg5, we can have %arg5 = ConstantOp() {tensor_value = [0, 2]} when the function (model) is cloned for a set of parameter. Then we can use the existing methods to handle it, no need to check the special mapping.

We already know the input is constant because it is stored in the ONNX model. This is why we need perform the extraction of the constant values when we translate from ONNX to MLIR. This patch does exactly what you propose but it does it only if certain conditions are met. For example, when the constant argument is the second parameter of the Reshape operation (there could be other cases where we want to do the same). What we don't want to do is emit all constant arguments as ConstantOps indiscriminately. If we do that we end up pulling constant inputs such as the weights of the model which are better left to be read in when the model is invoked.

**Comment by doru1004** [comment]
> In other words, the solution for reshape with constant second operand should not be in frontend importer. It should be independent from the solution of parameter passing for model.

We allow the parameter to be transformed into a ConstantOp precisely because it is the second argument of a Reshape. We do not allow the transformation of constant input args into ConstantOps for any other cases (at least for now, we will of course expand this to other cases). The idea is that we want to control which cases allow this transformation to happen.

File: src/dialect/onnx/onnx_ops.cpp [hunk 1]
```diff
@@ -656,9 +656,27 @@ void ONNXReshapeOp::inferShapes() {
...
+  // resides in the `value` attribute of the `ConstantTensor` operation.
+  auto *secondArgDefiningOp = (*getODSOperands(1).begin()).getDefiningOp();
+  auto constantOp =
+      dyn_cast_or_null<mlir::ONNXConstantOp>(secondArgDefiningOp);
+
   SmallVector<int64_t, 2> dims;
-  for (int i = 0; i < outputRank; ++i)
-    dims.emplace_back(-1);
+  if (!constantOp) {
+    for (int i = 0; i < outputRank; ++i)
```
Comment (AlexandreEichenberger) [inline src/dialect/onnx/onnx_ops.cpp]: you should use the method that pushes "n" constant values "c"

**Comment by doru1004** [comment]
Resolved comments.

**Comment by tungld** [comment]
@doru1004 Sorry for commenting in this closed patch. I just realized that with this patch, `numInputs` in `"onnx.EntryPoint"` in mnist becomes weird. 

> This patch enables the emission of constant values to be passed to operations. The current patch applies this functionality to the second argument of the Reshape operation.
> 
> The outcome is the shape dimensions for Reshape part of the MNIST model are now known:
> 
> ```
> module {
>   func @main_graph(%arg0: tensor<1x1x28x28xf32>, %arg1: tensor<8x1x5x5xf32>, %arg2: tensor<8x1x1xf32>, %arg3: tensor<16x8x5x5xf32>, %arg4: tensor<16x1x1xf32>, %arg5: tensor<2xi64>, %arg6: tensor<16x4x4x10xf32>, %arg7: tensor<2xi64>, %arg8: tensor<1x10xf32>) -> tensor<1x10xf32> {
>     %0 = "onnx.ConstantTensor"() {value = [256, 10]} : () -> tensor<2xi64>
>     %1 = "onnx.Reshape"(%arg6, %0) : (tensor<16x4x4x10xf32>, tensor<2xi64>) -> tensor<256x10xf32>
>     %2 = "onnx.ConvNoBias"(%arg0, %arg1) {auto_pad = "SAME_UPPER", dilations = [1, 1], group = 1 : i64, kernel_shape = [5, 5], strides = [1, 1]} : (tensor<1x1x28x28xf32>, tensor<8x1x5x5xf32>) -> tensor<1x8x28x28xf32>
>     %3 = "onnx.Add"(%2, %arg2) : (tensor<1x8x28x28xf32>, tensor<8x1x1xf32>) -> tensor<1x8x28x28xf32>
>     %4 = "onnx.Relu"(%3) : (tensor<1x8x28x28xf32>) -> tensor<1x8x28x28xf32>
>     %5 = "onnx.MaxPoolSingleOut"(%4) {auto_pad = "NOTSET", dilations = [1, 1], kernel_shape = [2, 2], pads = [0, 0, 0, 0], strides = [2, 2]} : (tensor<1x8x28x28xf32>) -> tensor<1x8x14x14xf32>
>     %6 = "onnx.ConvNoBias"(%5, %arg3) {auto_pad = "SAME_UPPER", dilations = [1, 1], group = 1 : i64, kernel_shape = [5, 5], strides = [1, 1]} : (tensor<1x8x14x14xf32>, tensor<16x8x5x5xf32>) -> tensor<1x16x14x14xf32>
>     %7 = "onnx.Add"(%6, %arg4) : (tensor<1x16x14x14xf32>, tensor<16x1x1xf32>) -> tensor<1x16x14x14xf32>
>     %8 = "onnx.Relu"(%7) : (tensor<1x16x14x14xf32>) -> tensor<1x16x14x14xf32>
>     %9 = "onnx.MaxPoolSingleOut"(%8) {auto_pad = "NOTSET", dilations = [1, 1], kernel_shape = [3, 3], pads = [0, 0, 0, 0], strides = [3, 3]} : (tensor<1x16x14x14xf32>) -> tensor<1x16x4x4xf32>
>     %10 = "onnx.ConstantTensor"() {value = [1, 256]} : () -> tensor<2xi64>
>     %11 = "onnx.Reshape"(%9, %10) : (tensor<1x16x4x4xf32>, tensor<2xi64>) -> tensor<1x256xf32>
>     %12 = "onnx.Gemm"(%11, %1, %arg8) {alpha = 1.000000e+00 : f32, beta = 1.000000e+00 : f32, transA = 0 : i64, transB = 0 : i64} : (tensor<1x256xf32>, tensor<256x10xf32>, tensor<1x10xf32>) -> tensor<1x10xf32>
>     return %12 : tensor<1x10xf32>
>   }
>   "onnx.EntryPoint"() {func = @main_graph, numInputs = 1 : i32, numOutputs = 1 : i32} : () -> ()
> }
> ```

Here `numInputs = 1` while we have 9 arguments in the `main_graph` function?

**Pull request merged**